### PR TITLE
Add black code formatter with pre-commit hook

### PR DESCRIPTION
**Pull request by jessesnyder**
## Description
- Reformats the codebase using the _black_ code formatter
- Adds an installable pre-commit hook which will run _black_ on commit
- Updates flake8 configuration to work with _black_
- Runs _flake8_ on `git push`

## Motivation and Context
The main reason to adopt it isn't necessarily to improve code format (since that's subjective), but to enforce stylistic consistency _without having to think about it_: it removes the burden of making judgements about how to format code. The _black_ project intro paragraph:

> By using Black, you agree to cede control over minutiae of hand-formatting. In return, Black gives you speed, determinism, and freedom from pycodestyle nagging about formatting. You will save time and mental energy for more important matters.
> Black makes code review faster by producing the smallest diffs possible. Blackened code looks the same regardless of the project you’re reading. Formatting becomes transparent after a while and you can focus on the content instead.

## How Has This Been Tested?
- Safety of code updates tested with existing test suite
- Pre-commit hook installation tested manually

### CODE REVIEW

**Comment by witekdev** [comment]
Interesting. Looks like sometimes it makes the Dallinger code more compact and other times more verbose. Not sure if I'm a fan of the expansion of 'complex' if statements:
in ``` dallinger/config.py ```
```
if extra_parameters is not None and getattr(extra_parameters, 'loaded', None) is None:

if (
	extra_parameters is not None
	and getattr(extra_parameters, "loaded", None) is None
):
```

**Comment by alecpm** [comment]
Seems to be primarily changing `'` to `"` which I'm neutral on, and changes to line breaking based on a consistent max line length which I generally like, though stuff like this trailing `):` makes me cringe a bit:

    def restore_state_from_replay(
        self, app_id, session, zip_path=None, **configuration_options
    ):

It's OK with me if people feel positively about it.

**Comment by witekdev** [comment]
@jessesnyder What was the rationale for experimenting with using this?
I'm also neutral however I do not see the clear advantages?
To me it seems like it makes certain things more compact and others more verbose, all seems a bit arbirary? :thinking:

**Comment by jessesnyder** [comment]
I think the main reason to adopt it wouldn't be to improve code format (since that's subjective), but to enforce stylistic consistency _without having to think about it_. It removes needing to make judgements about how to format, and for me, that's valuable. Their intro paragraph makes this case more articulately:

> By using Black, you agree to cede control over minutiae of hand-formatting. In return, Black gives you speed, determinism, and freedom from pycodestyle nagging about formatting. You will save time and mental energy for more important matters.
> 
> Black makes code review faster by producing the smallest diffs possible. Blackened code looks the same regardless of the project you’re reading. Formatting becomes transparent after a while and you can focus on the content instead.

**Comment by vlall** [comment]
What's needed before we merge this?

**Comment by jessesnyder** [comment]
@vlall I think some review of the git commit hook is warranted, so we'll probably need to defer merge until next iteration.

**Comment by jessesnyder** [comment]
@vlall @witekdev So I've made things confusing by revising the intent from "just preview the changes _black_ makes" to "actually make the commit hook and apply the changes". For PR review, I'd focus on the instructions for installing the pre-commit hook (which will require a `pip install -r dev-requirements` to get the `precommit` package) and running that. This probably means creating a branch _off of this branch_ so you can commit changes you know you'll throw away (since testing what happens on commit is the point).

**Comment by vlall** [comment]
So i created a branch called `stories/466-blackened-test`, installed the requirements using `pip install -r dev-requirements.txt`, and pushed a commit to the new branch. Everything looks good.

**Comment by vlall** [comment]
Just needs a rebase

**Comment by vlall** [comment]
Where are the "instructions" you mention that I should look at?

File: docs/source/developing_dallinger_setup_guide.rst [hunk 1]
```diff
@@ -680,6 +680,21 @@ Test that your installation works by running:
     dallinger --version
 
 
+Install the Git pre-commit hook
```
Comment (jessesnyder) [inline docs/source/developing_dallinger_setup_guide.rst]: @vlall Here are the instructions on installing the pre-commit hook. Once installed, it should run `black` on commit.

**Comment by vlall** [comment]
I'm actually concerned about this now... shouldn't this have been formatted after i made the commit? https://github.com/Dallinger/Dallinger/commit/5773000483634593485e1410fb3149d51b7669d6

File: docs/source/developing_dallinger_setup_guide.rst (hunk 1 above)
Comment (vlall) [reply docs/source/developing_dallinger_setup_guide.rst]: The instructions seem fine, do I need to do anything differently when making a commit to get the contents of my commit to format? It should just auto-format my usual commit, right?

**Comment by vlall** [comment]
This is now working with my last commit test thanks to @jessesnyder's fixes... I think it'd be useful to have a second pair of eyes after the latest changes made given how big this is.

**Pull request merged**